core/drivers/atmel_trng.c

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2021 Microchip
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <matrix.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+
+/* Registers */
+#define TRNG_CTRL		0x0
+#define TRNG_CTRL_WAKEY_OFFSET	8
+#define TRNG_CTRL_WAKEY_VALUE	0x524E47
+
+#define TRNG_IER		0x10
+#define TRNG_ISR		0x1C
+#define TRNG_ODATA		0x50
+
+static unsigned int trng_lock = SPINLOCK_UNLOCK;
+static unsigned int trng_random_val_lock = SPINLOCK_UNLOCK;
+static vaddr_t trng_base;
+static uint8_t random_byte_pos;
+static union {
+	uint32_t val;
+	uint8_t byte[sizeof(uint32_t)];
+} random_data;
+
+static uint32_t atmel_trng_read32(void)
+{
+	uint32_t exceptions = 0;
+	uint32_t value = 0;
+
+	exceptions = cpu_spin_lock_xsave(&trng_lock);
+
+	while (!io_read32(trng_base + TRNG_ISR))
+		;
+
+	value = io_read32(trng_base + TRNG_ODATA);
+
+	cpu_spin_unlock_xrestore(&trng_lock, exceptions);
+
+	return value;
+}
+
+TEE_Result crypto_rng_read(void *buf, size_t len)
+{
+	uint8_t *rngbuf = buf;
+	uint32_t val = 0;
+	uint8_t len_to_copy = 0;
```

Comment:
(late minor comment) no gain in specifically using `uint8_t` here, `size_t` would do the job.